File: fxgl/src/main/java/com/almasb/fxgl/app/FXGLApplication.java
/*
 * The MIT License (MIT)
 *
 * FXGL - JavaFX Game Library
 *
 * Copyright (c) 2015-2017 AlmasB (almaslvl@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.almasb.fxgl.app;

import com.almasb.fxgl.core.concurrent.Async;
import com.almasb.fxgl.core.logging.FXGLLogger;
import com.almasb.fxgl.core.logging.Logger;
import com.almasb.fxgl.gameplay.AchievementManager;
import com.almasb.fxgl.scene.PreloadingScene;
import com.almasb.fxgl.service.*;
import com.almasb.fxgl.settings.GameSettings;
import com.almasb.fxgl.settings.ReadOnlyGameSettings;
import com.almasb.fxgl.util.Version;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.geometry.Rectangle2D;
import javafx.stage.Stage;
import javafx.stage.StageStyle;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.PropertyResourceBundle;
import java.util.ResourceBundle;

/**
 * General FXGL application that configures services, settings and properties
 * to be used by the framework.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public abstract class FXGLApplication extends Application {

    /**
     * Use system logger fallback until actual logger is ready.
     */
    protected static Logger log = FXGLLogger.getSystemLogger();

    private Stage primaryStage;

    /**
     * Used by mocking.
     *
     * @param stage mock stage
     */
    void injectStage(Stage stage) {
        primaryStage = stage;
    }

    /**
     * @return primary stage as set by JavaFX
     */
    public final Stage getPrimaryStage() {
        return primaryStage;
    }

    /**
     * Internal copy of settings, so that they are not modified during game.
     */
    private ReadOnlyGameSettings settings;

    /**
     * Used by mocking.
     *
     * @param settings mock settings
     */
    void injectSettings(ReadOnlyGameSettings settings) {
        this.settings = settings;
    }

    /**
     * @return read only copy of game settings
     */
    public final ReadOnlyGameSettings getSettings() {
        return settings;
    }

    private List<FXGLListener> systemListeners = new ArrayList<>();

    /**
     * Add listener for core FXGL callbacks.
     *
     * @param listener the listener
     */
    public final void addFXGLListener(FXGLListener listener) {
        systemListeners.add(listener);
    }

    /**
     * Remove previously added listener.
     *
     * @param listener the listener
     */
    public final void removeFXGLListener(FXGLListener listener) {
        systemListeners.remove(listener);
    }

    public void runTask(Class<? extends Runnable> type) {
        log.debug("Running task: " + type.getSimpleName());
        FXGL.getInstance(type).run();
    }

    @Override
    public final void init() throws Exception {
        Version.print();
    }

    @Override
    public final void start(Stage primaryStage) throws Exception {
        this.primaryStage = primaryStage;
        showPreloadingStage();
        startFXGL();
    }

    private void startFXGL() {
        new Thread(() -> {
            try {
                configureFXGL();

                runUpdaterAndWait();

                configureApp();
            } catch (Exception e) {
                log.fatal("Exception during system configuration:");
                log.fatal(FXGLLogger.errorTraceAsString(e));
                log.fatal("System will now exit");
                log.close();

                // we don't know what exactly has been initialized
                // so to avoid the process hanging just shut down the JVM
                System.exit(-1);
            }
        }, "FXGL Launcher Thread").start();
    }

    private void runUpdaterAndWait() {
        Async.startFX(() -> runTask(UpdaterTask.class)).await();
    }

    @Override
    public final void stop() {}

    /**
     * Shows preloading stage with scene while FXGL is being configured.
     */
    private void showPreloadingStage() {
        Stage preloadingStage = new Stage(StageStyle.UNDECORATED);
        preloadingStage.initOwner(primaryStage);
        preloadingStage.setScene(new PreloadingScene());
        preloadingStage.show();

        // when main stage is ready to show
        primaryStage.setOnShowing(e -> {
            // close our preloader
            preloadingStage.close();
            // clean the reference to lambda + preloader
            primaryStage.setOnShowing(null);
        });
    }

    /**
     * After this call all FXGL.* calls are valid.
     */
    private void configureFXGL() {
        long start = System.nanoTime();

        initSystemProperties();
        initUserProperties();
        initAppSettings();

        FXGL.configure(new ApplicationModule((GameApplication) this));

        log = FXGLLogger.get(FXGLApplication.class);
        log.debug("FXGL configuration complete");

        log.infof("FXGL configuration took:  %.3f sec", (System.nanoTime() - start) / 1000000000.0);

        log.debug("Logging game settings\n" + settings.toString());
    }

    /**
     * Configure the actual application that uses FXGL.
     * Called after the FXGL has been configured.
     */
    abstract void configureApp();

//    /**
//     * Pause execution of current state.
//     */
//    public void pause() {
//        log.debug("Pausing application");
//        systemListeners.forEach(FXGLListener::onPause);
//    }
//
//    /**
//     * Resume execution of current state.
//     */
//    public void resume() {
//        log.debug("Resuming application");
//        systemListeners.forEach(FXGLListener::onResume);
//    }
//
//    /**
//     * Reset the application.
//     * After notifying all interested parties (where all should do a cleanup),
//     * <code>System.gc()</code> will be called.
//     */
//    public final void reset() {
//        log.debug("Resetting FXGL application");
//        systemListeners.forEach(FXGLListener::onReset);
//
//        System.gc();
//    }

    /**
     * Exit the application.
     * Safe to call this from a paused state.
     */
    protected final void exit() {
        log.debug("Exiting FXGL application");
        systemListeners.forEach(FXGLListener::onExit);

        FXGL.destroy();
        log.debug("Closing FXGL logger and exiting JavaFX");
        log.close();

        Platform.exit();
    }

    /**
     * Load FXGL system properties.
     */
    private void initSystemProperties() {
        ResourceBundle props = ResourceBundle.getBundle("com.almasb.fxgl.app.system");
        props.keySet().forEach(key -> {
            Object value = props.getObject(key);
            FXGL.setProperty(key, value);
        });
    }

    /**
     * Load user defined properties to override FXGL system properties.
     */
    private void initUserProperties() {
        // services are not ready yet, so load manually
        try (InputStream is = getClass().getResource("/assets/properties/system.properties").openStream()) {
            ResourceBundle props = new PropertyResourceBundle(is);
            props.keySet().forEach(key -> {
                Object value = props.getObject(key);
                FXGL.setProperty(key, value);
            });
        } catch (NullPointerException npe) {
            // User properties not found. Using system
        } catch (IOException e) {
            log.warning("Loading user properties failed: " + e);
        }
    }

    /**
     * Take app settings from user.
     */
    private void initAppSettings() {
        GameSettings localSettings = new GameSettings();
        initSettings(localSettings);
        settings = localSettings.toReadOnly();
    }

    /**
     * Initialize app settings.
     *
     * @param settings app settings
     */
    protected abstract void initSettings(GameSettings settings);

    /**
     * @return target width as set by GameSettings
     */
    public final int getWidth() {
        return getSettings().getWidth();
    }

    /**
     * @return target height as set by GameSettings
     */
    public final int getHeight() {
        return getSettings().getHeight();
    }

    /**
     * @return app bounds as set by GameSettings
     * @apiNote equivalent to new Rectangle2D(0, 0, getWidth(), getHeight())
     */
    public final Rectangle2D getAppBounds() {
        return new Rectangle2D(0, 0, getWidth(), getHeight());
    }

    /**
     * @return current tick (frame)
     */
    public final long getTick() {
        return 0;
    }

    /**
     * @return current time since start of game in nanoseconds
     */
    public final long getNow() {
        return 0;
    }

    public final EventBus getEventBus() {
        return FXGL.getEventBus();
    }

    public final Display getDisplay() {
        return FXGL.getDisplay();
    }

    public Input getInput() {
        return FXGL.getInput();
    }

    public final AudioPlayer getAudioPlayer() {
        return FXGL.getAudioPlayer();
    }

    public final AssetLoader getAssetLoader() {
        return FXGL.getAssetLoader();
    }

    public StateTimer getMasterTimer() {
        return FXGL.getMasterTimer();
    }

    public final Executor getExecutor() {
        return FXGL.getExecutor();
    }

    public final NotificationService getNotificationService() {
        return FXGL.getNotificationService();
    }

    public final AchievementManager getAchievementManager() {
        return FXGL.getAchievementManager();
    }

    public final QTE getQTE() {
        return FXGL.getQTE();
    }

    public final Net getNet() {
        return FXGL.getNet();
    }

    public final ExceptionHandler getExceptionHandler() {
        return FXGL.getExceptionHandler();
    }

    public final UIFactory getUIFactory() {
        return FXGL.getUIFactory();
    }

    public final QuestService getQuestService() {
        return FXGL.getQuestManager();
    }
}

File: fxgl/src/main/kotlin/com/almasb/fxgl/app/FXGLApplication.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.app

import com.almasb.fxgl.animation.Interpolators
import com.almasb.fxgl.app.scene.FXGLScene
import com.almasb.fxgl.app.scene.GameScene
import com.almasb.fxgl.app.scene.LoadingScene
import com.almasb.fxgl.app.services.FXGLAssetLoaderService
import com.almasb.fxgl.core.Updatable
import com.almasb.fxgl.core.collection.PropertyMap
import com.almasb.fxgl.core.concurrent.Async
import com.almasb.fxgl.core.concurrent.IOTask
import com.almasb.fxgl.core.serialization.Bundle
import com.almasb.fxgl.core.util.PauseMenuBGGen
import com.almasb.fxgl.dsl.FXGL
import com.almasb.fxgl.dsl.animationBuilder
import com.almasb.fxgl.dsl.getGameController
import com.almasb.fxgl.entity.GameWorld
import com.almasb.fxgl.event.EventBus
import com.almasb.fxgl.input.Input
import com.almasb.fxgl.input.InputSequence
import com.almasb.fxgl.input.UserAction
import com.almasb.fxgl.localization.LocalizationService
import com.almasb.fxgl.logging.Logger
import com.almasb.fxgl.physics.PhysicsWorld
import com.almasb.fxgl.profile.DataFile
import com.almasb.fxgl.profile.SaveLoadHandler
import com.almasb.fxgl.profile.SaveLoadService
import com.almasb.fxgl.scene.Scene
import com.almasb.fxgl.scene.SceneService
import com.almasb.fxgl.scene.SubScene
import com.almasb.fxgl.texture.toBufferedImage
import com.almasb.fxgl.texture.toImage
import com.almasb.fxgl.time.Timer
import com.almasb.fxgl.ui.DialogService
import com.almasb.fxgl.ui.FontType
import com.gluonhq.attach.lifecycle.LifecycleEvent
import com.gluonhq.attach.lifecycle.LifecycleService
import javafx.application.Application
import javafx.beans.property.ReadOnlyDoubleProperty
import javafx.beans.property.SimpleDoubleProperty
import javafx.concurrent.Task
import javafx.event.EventHandler
import javafx.scene.Group
import javafx.scene.ImageCursor
import javafx.scene.image.ImageView
import javafx.scene.input.KeyCode.*
import javafx.scene.input.KeyEvent
import javafx.scene.input.MouseEvent
import javafx.scene.paint.Color
import javafx.scene.shape.Circle
import javafx.stage.Stage
import javafx.util.Duration
import java.nio.file.Files
import java.nio.file.Paths
import java.time.LocalDateTime
import java.util.*
import javax.imageio.ImageIO
import kotlin.collections.HashMap
import kotlin.system.measureNanoTime

/**
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */

class FXGLApplication : Application() {

    companion object {
        private val log = Logger.get(FXGLApplication::class.java)

        private lateinit var app: GameApplication
        private lateinit var settings: ReadOnlyGameSettings

        private lateinit var engine: Engine
        private lateinit var mainWindow: MainWindow

        @JvmStatic fun launchFX(app: GameApplication, settings: ReadOnlyGameSettings, args: Array<String>) {
            this.app = app
            this.settings = settings

            Application.launch(FXGLApplication::class.java, *args)
        }

        @JvmStatic fun embeddedLaunchFX(app: GameApplication, settings: ReadOnlyGameSettings): FXGLPane {
            this.app = app
            this.settings = settings

            return FXGLApplication().embeddedStart()
        }
    }

    /**
     * This is the main entry point when run inside an existing JavaFX application.
     */
    private fun embeddedStart(): FXGLPane {
        val pane = FXGLPane(settings.width.toDouble(), settings.height.toDouble())

        startImpl {
            EmbeddedPaneWindow(pane, settings.sceneFactory.newStartup(settings.width, settings.height), settings)
        }

        return pane
    }

    /**
     * This is the main entry point as run by the JavaFX platform.
     */
    override fun start(stage: Stage) {
        startImpl {
            PrimaryStageWindow(stage, settings.sceneFactory.newStartup(settings.width, settings.height), settings)
        }
    }

    private fun startImpl(windowSupplier: () -> MainWindow) {
        // any exception on the JavaFX thread will be caught and reported
        Thread.setDefaultUncaughtExceptionHandler { _, e ->
            handleFatalError(e)
        }

        log.debug("Initializing FXGL")

        engine = Engine(settings)

        // after this call, all FXGL.* calls (apart from those accessing services) are valid
        FXGL.inject(engine, app, this)

        // get window up ASAP
        mainWindow = windowSupplier()
        mainWindow.show()

        // start initialization of services on a background thread
        // then start the loop on the JavaFX thread
        val task = IOTask.ofVoid {
            val time = measureNanoTime {
                engine.initServices()
                postServicesInit()
            }

            log.infof("FXGL initialization took: %.3f sec", time / 1000000000.0)
        }
                .onSuccess {
                    engine.startLoop()
                    setFirstSceneAfterStartup()
                }
                .onFailure { handleFatalError(it) }
                .toJavaFXTask()

        Async.execute(task)
    }

    private fun postServicesInit() {
        // fonts take a (relatively) long time to load, so load them in parallel
        Async.startAsync {
            log.debug("Loading fonts")

            val uiFactory = FXGL.getUIFactoryService()

            val fontUI = FXGL.getAssetLoader().loadFont(settings.fontUI)
            val fontGame = FXGL.getAssetLoader().loadFont(settings.fontGame)
            val fontMono = FXGL.getAssetLoader().loadFont(settings.fontMono)
            val fontText = FXGL.getAssetLoader().loadFont(settings.fontText)

            // but register them on the JavaFX thread
            Async.startAsyncFX {
                log.debug("Registering font factories with UI factory")

                uiFactory.registerFontFactory(FontType.UI, fontUI)
                uiFactory.registerFontFactory(FontType.GAME, fontGame)
                uiFactory.registerFontFactory(FontType.MONO, fontMono)
                uiFactory.registerFontFactory(FontType.TEXT, fontText)
            }
        }

        initPauseResumeHandler()
        initSaveLoadHandler()
        initAndLoadLocalization()

        // onGameUpdate is only updated in Game Scene
        FXGL.getGameScene().addListener(Updatable { tpf -> engine.onGameUpdate(tpf) })
    }

    private fun initPauseResumeHandler() {
        if (settings.isMobile) {
            initPauseResumeHandlerMobile()
        } else {
            initPauseResumeHandlerDesktop()
        }
    }

    private fun initPauseResumeHandlerMobile() {
        val serviceWrapper = LifecycleService.create()

        if (serviceWrapper.isEmpty) {
            log.warning("Attach LifecycleService is not present")
        } else {
            val service = serviceWrapper.get()

            log.debug("Init pause/resume handlers via Attach LifecycleService")

            service.addListener(LifecycleEvent.PAUSE) {
                engine.pauseLoop()
            }

            service.addListener(LifecycleEvent.RESUME) {
                engine.resumeLoop()
            }
        }
    }

    private fun initPauseResumeHandlerDesktop() {
        mainWindow.iconifiedProperty().addListener { _, _, isMinimized ->
            if (isMinimized) {
                engine.pauseLoop()
            } else {
                engine.resumeLoop()
            }
        }
    }

    private fun initSaveLoadHandler() {
        FXGL.getSaveLoadService().addHandler(object : SaveLoadHandler {
            override fun onSave(data: DataFile) {
                val bundle = Bundle("FXGLServices")
                engine.write(bundle)

                data.putBundle(bundle)
            }

            override fun onLoad(data: DataFile) {
                val bundle = data.getBundle("FXGLServices")
                engine.read(bundle)
            }
        })
    }

    private fun initAndLoadLocalization() {
        log.debug("Loading default localization")

        val defaultLang = settings.language.value

        val langData = FXGL.getAssetLoader().loadPropertyMap("languages/" + defaultLang.name.lowercase() + ".lang")

        FXGL.getLocalizationService().addLanguageData(defaultLang, langData.toStringMap())

        settings.supportedLanguages.filter { it != defaultLang }.forEach { lang ->
            FXGL.getLocalizationService().addLanguageDataLazy(lang) {
                FXGL.getAssetLoader()
                        .loadPropertyMap("languages/" + lang.name.lowercase() + ".lang")
                        .toStringMap()
            }
        }

        FXGL.getLocalizationService().selectedLanguageProperty().bind(settings.language)
    }

    private fun setFirstSceneAfterStartup() {
        // Start -> (Intro) -> (Menu) -> Game
        if (settings.isIntroEnabled) {
            getGameController().gotoIntro()
        } else {
            if (settings.isMainMenuEnabled) {
                getGameController().gotoMainMenu()
            } else {
                getGameController().startNewGame()
            }
        }
    }

    private var isError = false

    private fun handleFatalError(error: Throwable) {
        if (isError) {
            // just ignore to avoid spamming dialogs
            return
        }

        isError = true

        // stop main loop from running as we cannot continue
        engine.stopLoop()

        log.fatal("Uncaught Exception:", error)
        log.fatal("Application will now exit")

        mainWindow.showFatalError(error, Runnable {
            exitFXGL()
        })
    }

    fun exitFXGL() {
        log.debug("Exiting FXGL")

        if (!isError) {
            engine.stopLoopAndExitServices()
            app.onExit()
        }

        Async.shutdownNow()

        mainWindow.close()

        Logger.close()
    }

    /**
     * Clears previous game.
     * Initializes game, physics and UI.
     * This task is rerun every time the game application is restarted.
     */
    class InitAppTask : Task<Void?>() {

        override fun call(): Void? {
            val time = measureNanoTime {

                log.debug("Initializing game");
                updateMessage("Initializing game");

                initGame();
                app.initPhysics();
                app.initUI();

                engine.onGameReady(FXGL.getWorldProperties());
            }

            log.infof("Game initialization took: %.3f sec", time / 1000000000.0);

            return null
        }

        private fun initGame() {
            val vars = HashMap<String, Any>()
            app.initGameVars(vars)

            vars.forEach { name, value ->
                FXGL.getWorldProperties().setValue(name, value)
            }

            engine.onVarsInitialized(FXGL.getWorldProperties())

            app.initGame();
        }

        override fun failed() {
            throw RuntimeException("Initialization failed", exception)
        }
    }

    class GameApplicationService : SceneService() {

        private val log = Logger.get(javaClass)

        private lateinit var assetLoaderService: FXGLAssetLoaderService
        private lateinit var saveLoadService: SaveLoadService

        private lateinit var localService: LocalizationService
        private lateinit var dialogService: DialogService

        /**
         * The root for the overlay group that is constantly visible and on top
         * of every other UI element. For things like notifications.
         */
        override val overlayRoot = Group()

        override fun prefWidthProperty(): ReadOnlyDoubleProperty = settings.prefWidthProperty()

        override fun prefHeightProperty(): ReadOnlyDoubleProperty = settings.prefHeightProperty()

        override val eventBus = EventBus()

        override val input: Input
            get() = mainWindow.input

        /**
         * Always-on timer.
         */
        override val timer = Timer()

        override val worldProperties: PropertyMap
            get() = gameScene.gameWorld.properties

        override val currentScene: Scene
            get() = mainWindow.currentScene

        private lateinit var gameSceneRef: GameScene
        private lateinit var loadScene: LoadingScene

        private var intro: FXGLScene? = null
        private var mainMenu: SubScene? = null
        private var gameMenu: SubScene? = null

        override val introScene: Optional<Scene>
            get() = Optional.ofNullable(intro)

        override val loadingScene: Optional<Scene>
            get() = Optional.ofNullable(loadScene)

        override val mainMenuScene: Optional<Scene>
            get() = Optional.ofNullable(mainMenu)

        override val gameMenuScene: Optional<Scene>
            get() = Optional.ofNullable(gameMenu)

        override val gameScene: GameScene
            get() = gameSceneRef

        internal val window: MainWindow
            get() = mainWindow

        override fun isInHierarchy(scene: Scene): Boolean {
            return mainWindow.isInHierarchy(scene)
        }

        override fun onInit() {
            settings.cssList.forEach {
                log.debug("Applying CSS: $it")
                mainWindow.addCSS(assetLoaderService.loadCSS(it))
            }

            mainWindow.onClose = {
                if (settings.isCloseConfirmation) {
                    if (canShowCloseDialog()) {
                        showConfirmExitDialog()
                    }
                } else {
                    FXGL.getGameController().exit()
                }
            }

            mainWindow.currentSceneProperty.addListener { _, oldScene, newScene ->
                log.debug("Removing overlay from $oldScene and adding to $newScene")

                removeOverlay(oldScene)
                addOverlay(newScene)
            }

            initAppScenes()
        }

        override fun onGameUpdate(tpf: Double) {
            app.onUpdate(tpf)
        }

        override fun onUpdate(tpf: Double) {
            timer.update(tpf)
            mainWindow.update(tpf)
        }

        private fun initAppScenes() {
            log.debug("Initializing application scenes")

            val sceneFactory = settings.sceneFactory

            loadScene = sceneFactory.newLoadingScene()
            gameSceneRef = GameScene(settings.width, settings.height,
                    GameWorld(),
                    PhysicsWorld(settings.height, settings.pixelsPerMeter, settings.collisionDetectionStrategy),
                    settings.is3D
            )

            gameSceneRef.isSingleStep = settings.isSingleStep

            if (settings.isClickFeedbackEnabled) {
                addClickFeedbackHandler()
            }

            if (settings.isIntroEnabled) {
                intro = sceneFactory.newIntro()
                        .also {
                            it.onFinished = Runnable {
                                setSceneAfterIntro()
                            }
                        }
            }

            if (settings.isMainMenuEnabled) {
                mainMenu = sceneFactory.newMainMenu()
            }

            if (settings.isGameMenuEnabled) {
                gameMenu = sceneFactory.newGameMenu()

                val menuKeyHandler = object : EventHandler<KeyEvent> {
                    private var canSwitchGameMenu = true

                    private fun onMenuKey(pressed: Boolean) {
                        if (!pressed) {
                            canSwitchGameMenu = true
                            return
                        }

                        if (canSwitchGameMenu) {
                            // we only care if menu key was pressed in one of these states
                            if (mainWindow.currentScene === gameMenu) {
                                canSwitchGameMenu = false
                                popSubScene()

                            } else if (mainWindow.currentScene === gameSceneRef) {
                                canSwitchGameMenu = false
                                pushSubScene(gameMenu!!)
                            }
                        }
                    }

                    override fun handle(event: KeyEvent) {
                        if (event.code == settings.menuKey) {
                            onMenuKey(event.eventType == KeyEvent.KEY_PRESSED)
                        }
                    }
                }

                gameSceneRef.input.addEventHandler(KeyEvent.ANY, menuKeyHandler)
                gameMenu!!.input.addEventHandler(KeyEvent.ANY, menuKeyHandler)
            }

            log.debug("Application scenes initialized")
        }

        private fun setSceneAfterIntro() {
            if (settings.isMainMenuEnabled) {
                getGameController().gotoMainMenu()
            } else {
                getGameController().startNewGame()
            }
        }

        private fun addClickFeedbackHandler() {
            gameSceneRef.input.addEventHandler(MouseEvent.MOUSE_PRESSED, EventHandler {
                val circle = Circle(gameSceneRef.input.mouseXUI, gameSceneRef.input.mouseYUI, 5.0, null)
                circle.stroke = Color.GOLD
                circle.strokeWidth = 2.0
                circle.opacityProperty().bind(SimpleDoubleProperty(1.0).subtract(circle.radiusProperty().divide(35.0)))

                overlayRoot.children += circle

                animationBuilder()
                        .interpolator(Interpolators.SMOOTH.EASE_IN())
                        .onFinished(Runnable { overlayRoot.children -= circle })
                        .duration(Duration.seconds(0.33))
                        .animate(circle.radiusProperty())
                        .to(35.0)
                        .buildAndPlay()
            })
        }

        override fun onMainLoopStarting() {
            // these things need to be called early before the main loop
            // so that menus can correctly display input controls, etc.
            app.initInput()
            app.onPreInit()

            if (!settings.isNative) {
                mainWindow.addIcons(assetLoaderService.loadImage(settings.appIcon))

                val cursorInfo = settings.defaultCursor

                mainWindow.defaultCursor = ImageCursor(
                        assetLoaderService.loadImage(cursorInfo.imageName),
                        cursorInfo.hotspotX,
                        cursorInfo.hotspotY
                )
            }

            SystemActions.bind(mainWindow.input)

            mainWindow.input.addAction(object : UserAction("") {
                private val subScene = object : SubScene() {
                    private val view by lazy {
                        ImageView(PauseMenuBGGen.generate().toImage()).also {
                            it.scaleX = 4.0
                            it.scaleY = 4.0
                            it.translateX = prefWidth / 2.0
                            it.translateY = prefHeight / 2.0
                        }
                    }

                    override fun onCreate() {
                        contentRoot.children.setAll(view)

                        timer.runOnceAfter({ popSubScene() }, Duration.seconds(3.0))
                    }
                }
                override fun onActionBegin() {
                    pushSubScene(subScene)
                }
            }, InputSequence(F, X, G, L, A, L, M, A, S, B))
        }

        private fun addOverlay(scene: Scene) {
            if (scene is FXGLScene) {
                scene.contentRoot.children += overlayRoot
            } else {
                scene.root.children += overlayRoot
            }
        }

        private fun removeOverlay(scene: Scene) {
            if (scene is FXGLScene) {
                scene.contentRoot.children -= overlayRoot
            } else {
                scene.root.children -= overlayRoot
            }
        }

        /**
         * @return true if can show close dialog
         */
        private fun canShowCloseDialog(): Boolean {
            // do not allow close dialog if
            // 1. a dialog is shown
            // 2. we are loading a game
            // 3. we are showing intro

            // mainWindow.currentScene === dialogScene ||
            val isNotOK =  mainWindow.currentScene === loadScene
                    || (settings.isIntroEnabled && mainWindow.currentScene === intro)

            return !isNotOK
        }

        private fun showConfirmExitDialog() {
            dialogService.showConfirmationBox(localService.getLocalizedString("dialog.exitGame")) { yes ->
                if (yes)
                    FXGL.getGameController().exit()
            }
        }

        fun startNewGame() {
            log.debug("Starting new game")

            mainWindow.setScene(loadScene)

            clearPreviousGame()

            loadScene.pushNewTask(InitAppTask())
        }

        private fun clearPreviousGame() {
            log.debug("Clearing previous game")

            gameSceneRef.reset()
            engine.resetServices()
        }

        fun saveGame(dataFile: DataFile) {
            saveLoadService.save(dataFile)
        }

        fun loadGame(dataFile: DataFile) {
            log.debug("Starting loaded game")
            mainWindow.setScene(loadScene)

            clearPreviousGame()

            loadScene.pushNewTask(Runnable {
                InitAppTask().run()
                saveLoadService.load(dataFile)
            })
        }

        fun gotoIntro() {
            mainWindow.setScene(intro!!)
        }

        private val dummyScene by lazy {
            object : FXGLScene() {
                override fun toString(): String {
                    return "FXGLMainMenuDummyScene"
                }
            }
        }

        fun gotoMainMenu() {
            if (!settings.isMainMenuEnabled) {
                log.warning("Ignoring gotoMainMenu() because main menu is not enabled")
                return
            }

            // since mainMenu is a subscene we need an actual scene before it
            mainWindow.setScene(dummyScene)
            mainWindow.pushState(mainMenu!!)
        }

        fun gotoGameMenu() {
            if (!settings.isGameMenuEnabled) {
                log.warning("Ignoring gotoGameMenu() because game menu is not enabled")
                return
            }

            // current scene should be gameScene
            mainWindow.pushState(gameMenu!!)
        }

        fun gotoLoading(loadingTask: Runnable) {
            loadScene.pushNewTask(loadingTask)
            mainWindow.setScene(loadScene)
        }

        fun gotoLoading(loadingTask: Task<*>) {
            loadScene.pushNewTask(loadingTask)
            mainWindow.setScene(loadScene)
        }

        fun gotoPlay() {
            mainWindow.setScene(gameSceneRef)
        }

        /**
         * Saves a screenshot of the current scene into a ".png" file,
         * named by title + version + time.
         */
        fun saveScreenshot(): Boolean {
            val fxImage = mainWindow.takeScreenshot()
            val img = toBufferedImage(fxImage)

            var fileName = "./" + settings.title + settings.version + LocalDateTime.now()
            fileName = fileName.replace(":", "_")

            try {
                val name = if (fileName.endsWith(".png")) fileName else "$fileName.png"

                Files.newOutputStream(Paths.get(name)).use {
                    return ImageIO.write(img, "png", it)
                }
            } catch (e: Exception) {
                log.warning("saveScreenshot($fileName.png) failed: $e")
                return false
            }
        }

        override fun pushSubScene(subScene: SubScene) {
            mainWindow.pushState(subScene)
        }

        override fun popSubScene() {
            mainWindow.popState()
        }
    }
}



